src/librbd/mirror/snapshot/GroupUnlinkPeerRequest.cc

```diff
@@ -0,0 +1,340 @@
+	return;
+      }
+    }
+  }
+
+  for (auto peer: *m_mirror_peer_uuids){
+    for (auto it = m_group_snaps.begin(); it != m_group_snaps.end(); it++) {
+      auto ns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
+	  &it->snapshot_namespace);
+      if (ns == nullptr) {
+	continue;
+      }
+      // FIXME: after relocate, on new primary the previous primary demoted
+      // snap is not getting deleted, until the next demotion.
+      if (ns->state != cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+	// Reset the count if the group was demoted.
+	count = 0;
+	continue;
+      }
+
+      if (ns->mirror_peer_uuids.count(peer) == 0) { 
+	continue;
+      }
+
+      count++;
+      if (count == m_max_snaps) {
+	unlink_snap = it;
+      }
+      if (count > m_max_snaps) {
+	process_snapshot(*unlink_snap, peer);
+	return;
+      }
+    }
+  }
+
+  ldout(m_cct, 10)<< "no more snaps" << dendl;
+  finish(0);
+}
+
+
+template <typename I>
+void GroupUnlinkPeerRequest<I>::list_group_snaps() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = util::create_context_callback<
+    GroupUnlinkPeerRequest<I>,
+    &GroupUnlinkPeerRequest<I>::handle_list_group_snaps>(
+      this);
+
+  m_group_snaps.clear();
+  auto req = group::ListSnapshotsRequest<I>::create(
+    m_group_io_ctx, m_group_id, true, true, &m_group_snaps, ctx);
+
+  req->send();
+}
+
+
+
+template <typename I>
+void GroupUnlinkPeerRequest<I>::handle_list_group_snaps(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to list group snapshots of group ID '"
+                 << m_group_id << "': " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  unlink_peer();
+}
+
+template <typename I>
+void GroupUnlinkPeerRequest<I>::process_snapshot(cls::rbd::GroupSnapshot group_snap,
+                                                 std::string mirror_peer_uuid) {
+  ldout(m_cct, 10) << "snap id: " << group_snap.id << dendl;
+  bool found = false;
+  bool has_newer_mirror_snap = false;
+
+  for (auto it = m_group_snaps.begin(); it != m_group_snaps.end(); it++) {
+    if (it->id  == group_snap.id) {
+      found = true;
+    } else if (found) {
+      auto ns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
+	  &it->snapshot_namespace);
+      if (ns != nullptr) {
+	has_newer_mirror_snap = true;
+	break;
+      }
+    }
+  }
+
+  if (!found) {
+    ldout(m_cct, 15) << "missing snapshot: snap_id=" << group_snap.id << dendl;
+    finish(-ENOENT);
+    return;  
+  }
+
+  if (has_newer_mirror_snap) {
+    remove_group_snapshot(group_snap); 
+  } else {
+    remove_peer_uuid(group_snap, mirror_peer_uuid);
+  }
+}
+
+
+template <typename I>
+void GroupUnlinkPeerRequest<I>::remove_peer_uuid(
+                              cls::rbd::GroupSnapshot group_snap,
+                              std::string mirror_peer_uuid) {
+  ldout(m_cct, 10) << dendl;
+
+    lderr(m_cct) << "calling remove_peer_uuid" << dendl;
```

Comment:
Remove lderr above

Reply:
Those were added to try to capture the EINVAL err ListSnapshotsRequest - they can be ignored for now as they will be removed.